pkg/review/commit_review_status.go

```diff
@@ -192,12 +207,59 @@ func (fn *CommitApprovalDoFn) ProcessElement(ctx context.Context, commit Commit,
 	emit(commitReviewStatus)
 }
 
+// ProcessElement is a DoFn implementation that takes a CommitReviewStatus
+// and populates its breakGlassIssue field (if necessary) and then emits it
+// using the given emit function. The process only searches for break glass
+// issues for commits that do not have the status GithubPRApproved.
+func (fn *BreakGlassIssueDoFn) ProcessElement(ctx context.Context, commitReviewStatus CommitReviewStatus, emit func(status CommitReviewStatus)) {
+	logger := logging.FromContext(ctx)
+	logger.InfoContext(ctx, "processing commitReviewStatus", "commitReviewStatus", commitReviewStatus)
+	if commitReviewStatus.ApprovalStatus != GithubPRApproved {
+		// if the commit does not have proper approval, we check if there was a
+		// break glass issue opened by the author during the timeframe they
+		// submitted the commit.
+		breakGlassIssue, err := getBreakGlassIssue(
+			ctx,
+			fn.Client,
+			fn.Config.IssuesTable,
+			commitReviewStatus.Author,
+			commitReviewStatus.Timestamp,
+		)
+		if err != nil {
+			// We should only get transient style errors from BigQuery
+			// (e.g. network is down etc.). So, we can just log the error and then
+			// drop this CommitReviewStatus from the pipeline. It will then get
+			// retried on the next run of the pipeline.
+			logger.ErrorContext(ctx, "failure when trying to get break glass issue: %v", err)
+			return
+		}
+		commitReviewStatus.BreakGlassURL = breakGlassIssue.HTMLURL
+	}
+	emit(commitReviewStatus)
+}
+
+// getBreakGlassIssue Searches BigQuery for a break glass issue that was opened
+// by the given author and whose created_at date was on or before the given
+// timestamp and whose closed_at date was on or after the given timestamp.
+func getBreakGlassIssue(ctx context.Context, client bq.Client[breakGlassIssue], issuesTable bigqueryio.QualifiedTableName, author, timestamp string) (*breakGlassIssue, error) {
+	issueQuery := GetBreakGlassIssueQuery(issuesTable, author, timestamp)
+	items, err := client.Query(ctx, issueQuery)
+	if err != nil {
+		return nil, fmt.Errorf("client.Query failed: %w", err)
+	}
+	var breakGlassIssue *breakGlassIssue
+	if len(items) > 0 {
+		breakGlassIssue = &items[0]
+	}
```

Comment:
Should we handle a situation where there's more than one result? Would that be an error or a "should never be possible"? If there can be multiple results, are you sure we want the first one and not the last? Can we guarantee the results are returned in the same order?

Reply:
There should really only ever be at most one result returned. While I think it may technically possible for someone to open a second break glass issue if they already have one open, I can't think of a reason why someone would even do that as they already would already have break glass permissions from the first break glass issue they opened.

For the pipeline code though, all we really care about is whether the author had break glass permissions or not during the time the commit was made. So if the author had multiple break glass issues open at the same time, then any one of the issues is sufficient to establish that they had break glass permissions. So just using the first one is fine.

Reply:
Hmm - I wonder if we should attach _all_ breakglass issues (slice instead of string) to the PR.

Reply:
It wouldn't change the pipeline architecture that much to accommodate storing more than one break glass issue if there happen to be more than one. I've updated the code accordingly.